protomap/proto.go

```diff
@@ -361,20 +362,100 @@ func resolvedPath(basePath, annotatedPath *gpb.Path) *gpb.Path {
+//
+// This option is typically used in conjunction with path compression where there
+// are some leaves that do not have corresponding fields.
+func IgnoreExtraPaths() *ignoreExtraPaths { return &ignoreExtraPaths{} }
+
+type ignoreExtraPaths struct{}
+
+// isUnmapOpt marks ignoreExtraPaths as an unmap option.
+func (*ignoreExtraPaths) isUnmapOpt() {}
+
+// ValuePathPrefix indicates that the values in the supplied map have a prefix which
+// is equal to the supplied path. The prefix plus each path in the vals map must be
+// equal to the absolute path for the supplied values.
+func ValuePathPrefix(path *gpb.Path) *valuePathPrefix {
+	return &valuePathPrefix{p: path}
+}
+
+type valuePathPrefix struct{ p *gpb.Path }
+
+// isUnmapOpt marks valuePathPrefix as an unmap option.
+func (*valuePathPrefix) isUnmapOpt() {}
+
+// ProtobufMessagePrefix specifies the path that the protobuf message supplied to ProtoFromPaths
+// makes up. This is used in cases where the message itself is not the root - and hence unmarshalling
+// should look for paths relative to the specified path in the vals map.
+func ProtobufMessagePrefix(path *gpb.Path) *protoMsgPrefix {
+	return &protoMsgPrefix{p: path}
+}
+
+type protoMsgPrefix struct{ p *gpb.Path }
+
+// isUnmapOpt marks protoMsgPrefix as an unmap option.
+func (*protoMsgPrefix) isUnmapOpt() {}
+
+// ProtoFromPaths takes an input ygot-generated protobuf and unmarshals the values provided in vals into the map.
+// The vals map must be keyed by the gNMI path to the leaf, with the interface{} value being the value that the
+// leaf at the field should be set to.
+//
+// The protobuf p is modified in place to add the values.
+//
+// The set of UnmapOpts that are provided (opt) are used to control the behaviour of unmarshalling the specified data.
+//
+// ProtoFromPaths returns an error if the data cannot be unmarshalled.
+func ProtoFromPaths(p proto.Message, vals map[*gpb.Path]interface{}, opt ...UnmapOpt) error {
 	if p == nil {
 		return errors.New("nil protobuf supplied")
 	}
 
+	valPrefix, err := hasValuePathPrefix(opt)
+	if err != nil {
+		return fmt.Errorf("invalid value prefix supplied, %v", err)
+	}
+
+	protoPrefix, err := hasProtoMsgPrefix(opt)
+	if err != nil {
+		return fmt.Errorf("invalid protobuf message prefix supplied in options, %v", err)
+	}
+
+	schemaPath := func(p *gpb.Path) *gpb.Path {
+		np := proto.Clone(p).(*gpb.Path)
+		for _, e := range np.Elem {
+			e.Key = nil
+		}
+		return np
+	}
+
+	// directCh is a map between the absolute schema path for a particular value, and
+	// the value specified.
 	directCh := map[*gpb.Path]interface{}{}
 	for p, v := range vals {
-		// TODO(robjs): needs fixing for compressed schemas.
-		if len(p.GetElem()) == len(prefix.GetElem())+1 {
-			directCh[p] = v
+		// make the path absolute, and a schema path.
+		pp := util.TrimGNMIPathElemPrefix(&gpb.Path{
+			Elem: append(append([]*gpb.PathElem{}, schemaPath(valPrefix).Elem...), p.Elem...),
+		}, protoPrefix)
```

Comment:
I actually don't think that we should - if someone gives us a path in `vals` that is not mapped then this might not be an error case - since we can ignore additional paths.